To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth your local device, follow [our guide](https://docs.unsloth.ai/get-started/install-and-update). This notebook is licensed [LGPL-3.0](https://github.com/unslothai/notebooks?tab=LGPL-3.0-1-ov-file#readme).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News


Introducing FP8 precision training for faster RL inference. [Read Blog](https://docs.unsloth.ai/new/fp8-reinforcement-learning).

Unsloth's [Docker image](https://hub.docker.com/r/unsloth/unsloth) is here! Start training with no setup & environment issues. [Read our Guide](https://docs.unsloth.ai/new/how-to-train-llms-with-unsloth-and-docker).

[gpt-oss RL](https://docs.unsloth.ai/new/gpt-oss-reinforcement-learning) is now supported with the fastest inference & lowest VRAM. Try our [new notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/gpt-oss-(20B)-GRPO.ipynb) which creates kernels!

Introducing [Vision](https://docs.unsloth.ai/new/vision-reinforcement-learning-vlm-rl) and [Standby](https://docs.unsloth.ai/basics/memory-efficient-rl) for RL! Train Qwen, Gemma etc. VLMs with GSPO - even faster with less VRAM.

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os, re
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    import torch; v = re.match(r"[0-9]{1,}\.[0-9]{1,}", str(torch.__version__)).group(0)
    xformers = "xformers==" + ("0.0.33.post1" if v=="2.9" else "0.0.32.post2" if v=="2.8" else "0.0.29.post3")
    !pip install --no-deps bitsandbytes accelerate {xformers} peft trl triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf "datasets==4.3.0" "huggingface_hub>=0.34.0" hf_transfer
    !pip install --no-deps unsloth
!pip install transformers==4.56.2
!pip install --no-deps trl==0.22.2

### Unsloth

In [2]:
from unsloth import FastVisionModel # FastLanguageModel for LLMs
import torch
from transformers import AutoModelForCausalLM ,AutoProcessor
# 4bit pre quantized models we support for 4x faster downloading + no OOMs.
fourbit_models = [
    "unsloth/Qwen3-VL-8B-Instruct-bnb-4bit", # Qwen 3 vision support
    "unsloth/Qwen3-VL-8B-Thinking-bnb-4bit",
    "unsloth/Qwen3-VL-32B-Instruct-bnb-4bit",
    "unsloth/Qwen3-VL-32B-Thinking-bnb-4bit",
] # More models at https://huggingface.co/unsloth

model_path = "unsloth/PaddleOCR-VL"
model, tokenizer = FastVisionModel.from_pretrained(
    model_path,
    max_seq_length = 1024, # Choose any for long context!
    load_in_4bit = False,     # 4bit uses much less memory
    load_in_8bit = False,    # A bit more accurate, uses 2x memory
    full_finetuning=True, # We support full finetuning now!
    auto_model=AutoModelForCausalLM,
    trust_remote_code = True,
    unsloth_force_compile = True,
)

🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.


2025-12-23 08:48:39.767162: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1766479720.135307      55 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1766479720.254305      55 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1766479721.071429      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1766479721.071464      55 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1766479721.071467      55 computation_placer.cc:177] computation placer alr

🦥 Unsloth Zoo will now patch everything to make training faster!


A new version of the following files was downloaded from https://huggingface.co/unsloth/PaddleOCR-VL:
- configuration_paddleocr_vl.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


Unsloth: WARNING `trust_remote_code` is True.
Are you certain you want to do remote code execution?
==((====))==  Unsloth 2025.12.8: Fast Paddleocr_Vl patching. Transformers: 4.56.2.
   \\   /|    Tesla T4. Num GPUs = 2. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.8.0+cu126. CUDA: 7.5. CUDA Toolkit: 12.6. Triton: 3.4.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.32.post2. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!
Unsloth: Float16 full finetuning uses more memory since we upcast weights to float32.


modeling_paddleocr_vl.py: 0.00B [00:00, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/unsloth/PaddleOCR-VL:
- modeling_paddleocr_vl.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


model.safetensors:   0%|          | 0.00/1.92G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/133 [00:00<?, ?B/s]

tokenizer_config.json: 0.00B [00:00, ?B/s]

tokenizer.model:   0%|          | 0.00/1.61M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.2M [00:00<?, ?B/s]

added_tokens.json: 0.00B [00:00, ?B/s]

special_tokens_map.json: 0.00B [00:00, ?B/s]

chat_template.jinja: 0.00B [00:00, ?B/s]

We now load the processor

In [3]:
processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True)

processor_config.json:   0%|          | 0.00/137 [00:00<?, ?B/s]

processing_paddleocr_vl.py: 0.00B [00:00, ?B/s]

A new version of the following files was downloaded from https://huggingface.co/unsloth/PaddleOCR-VL:
- processing_paddleocr_vl.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.


preprocessor_config.json:   0%|          | 0.00/641 [00:00<?, ?B/s]

image_processing_paddleocr_vl.py: 0.00B [00:00, ?B/s]

We now add LoRA adapters for parameter efficient finetuning - this allows us to only efficiently train 1% of all parameters.

**[NEW]** We also support finetuning ONLY the vision part of the model, or ONLY the language part. Or you can select both! You can also select to finetune the attention or the MLP layers!

In [4]:
model = FastVisionModel.get_peft_model(
    model,
    r = 16,
    lora_alpha = 16,
    lora_dropout = 0,
    bias = "none",
    random_state = 3407,
    use_rslora = False,
    target_modules = [
      "q_proj", "k_proj", "v_proj", "o_proj",


    ]
)

Unsloth: Full finetuning is enabled, so .get_peft_model has no effect


<a name="Data"></a>
### Data Prep
We'll be using a sampled dataset of handwritten maths formulas. The goal is to convert these images into a computer readable form - ie in LaTeX form, so we can render it. This can be very useful for complex formulas.

You can access the dataset [here](https://huggingface.co/datasets/unsloth/LaTeX_OCR). The full dataset is [here](https://huggingface.co/datasets/linxy/LaTeX_OCR).

We can also render the LaTeX in the browser directly!

To format the dataset, all vision finetuning tasks should be formatted as follows:

```python
[
{ "role": "user",
  "content": [{"type": "text",  "text": Q}, {"type": "image", "image": image} ]
},
{ "role": "assistant",
  "content": [{"type": "text",  "text": A} ]
},
]
```

Let's convert the dataset into the "correct" format for finetuning:

We look at how the conversations are structured for the first example:

In [5]:
import os
import pandas as pd
from PIL import Image
import io
from datasets import Dataset

# Define your paths
CSV_PATH =  "/kaggle/input/medical-prescription/dataset.csv"
IMAGES_FOLDER =  "/kaggle/input/medical-prescription/archive (8)/data"
def load_custom_data(csv_path, images_folder):
    df = pd.read_csv(csv_path, header=None)
    formatted_data = []
    print(f"Loading data from {csv_path}...")

    for index, row in df.iterrows():
        img_name = str(row[0])
        ground_truth = str(row[1])
        img_full_path = os.path.join(images_folder, img_name)
        try:
            image = Image.open(img_full_path).convert("RGB")
            formatted_data.append({"image": image, "text": ground_truth})
        except Exception as e:
            print(f"⚠️ Error loading image {img_name}: {e}")
            continue
    return formatted_data

# 1. Load Data
custom_raw_data = load_custom_data(CSV_PATH, IMAGES_FOLDER)

# 2. Convert to list format
instruction = "OCR:"
converted_dataset = []

for sample in custom_raw_data:
    conversation = [
        {
            "role": "user",
            "content": [
                # Text First, then Image
                {"type": "text", "text": instruction},
                {"type": "image", "image": sample["image"]}
            ]
        },
        {
            "role": "assistant",
            "content": [
                {"type": "text", "text": sample["text"]}
            ]
        }
    ]
    converted_dataset.append({ "images": [sample["image"]], "messages": conversation })

# 3. Create Dataset
hf_dataset = Dataset.from_list(converted_dataset)

# 4. === THE CORRECTED TRANSFORM ===
def format_output(batch):
    formatted_batch = {"images": [], "messages": []}

    for i in range(len(batch["images"])):
        # -- Step A: Prepare Master Image --
        raw_imgs = batch["images"][i]
        valid_images = []
        for img in raw_imgs:
            if isinstance(img, dict) and "bytes" in img:
                valid_images.append(Image.open(io.BytesIO(img["bytes"])).convert("RGB"))
            else:
                valid_images.append(img)

        formatted_batch["images"].append(valid_images)
        master_image = valid_images[0]

        # -- Step B: Fix Messages --
        raw_msgs = batch["messages"][i]
        clean_msgs = []

        for msg in raw_msgs:
            new_content = []
            for item in msg["content"]:
                # FIX: Check if image is present AND NOT NONE
                # The library inserts "image": None for text items, so we must check values.
                if "image" in item and item["image"] is not None:
                    new_content.append({"type": "image", "image": master_image})
                else:
                    # It's a text item. Clean up the None keys.
                    clean_item = {k: v for k, v in item.items() if v is not None}
                    new_content.append(clean_item)

            clean_msgs.append({"role": msg["role"], "content": new_content})

        formatted_batch["messages"].append(clean_msgs)

    return formatted_batch

# Apply the transform
hf_dataset.set_transform(format_output)

# 5. Split
dataset_split = hf_dataset.train_test_split(test_size=0.1)
train_dataset = dataset_split["train"]
eval_dataset= dataset_split["test"]

# === VERIFY OUTPUT ===
print("\nSample Output Structure:")
sample = train_dataset[0]

# Standard print allows you to see the true structure
print(sample)

Loading data from /kaggle/input/medical-prescription/dataset.csv...
⚠️ Error loading image Filename: [Errno 2] No such file or directory: '/kaggle/input/medical-prescription/archive (8)/data/Filename'

Sample Output Structure:
{'images': [<PIL.PngImagePlugin.PngImageFile image mode=RGB size=1024x1024 at 0x7F99448FCDD0>], 'messages': [{'role': 'user', 'content': [{'text': 'OCR:', 'type': 'text'}, {'type': 'image', 'image': <PIL.PngImagePlugin.PngImageFile image mode=RGB size=1024x1024 at 0x7F99448FCDD0>}]}, {'role': 'assistant', 'content': [{'text': 'Preethi Child Clinic Dr. D. Ravi Shankar Maruthi Complex, opp: South Indian bank, Suchitra. Ph: 9989102020 Timings: 9.00 AM to 10.00 am Name. 6.00 PM o 7.30 PM Shanaya M.D. (PED) PALS:MIAP Reg. No. 35400 In Case of Emergency Contact H.R. Children Hospital 10-57/A, Soubhagya Nagar, Opp: 1.D.PL Colony, Hyd Ph: 23078750 Timings: 10.00 AM to 1.00 PM 4.00 PM o 5.30 PM 8.00 PM to 10.00 PM Age montre Sex F wb.b14. Temp...........Date Cold2 con 2 d

In [6]:
from transformers import TextStreamer
def run_inference(idx, dataset):
    # 1. Setup Data
    # CHANGED: "image" -> "images"[0]
    image = dataset[idx]["images"][0]
    print(image)
    ground_truth = dataset[idx]["messages"][1]["content"][0]["text"]
    instruction = "OCR:"

    messages = [
        {"role": "user", "content": [
            {"type": "image"},
            {"type": "text", "text": instruction}
        ]}
    ]

    # 2. Prepare Inputs
    text_prompt = processor.tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )
    inputs = processor(
        image, text_prompt, add_special_tokens=False, return_tensors="pt"
    ).to("cuda")

    # 3. Generate
    print(f"\n--- Index {idx} ---")
    print(f"Ground Truth: {ground_truth}")
    print("Prediction: ", end="")

    streamer = TextStreamer(processor.tokenizer, skip_prompt=True)
    _ = model.generate(
        **inputs,
        streamer=streamer,
        max_new_tokens=128,
        use_cache=False,
        temperature=1.5,
        min_p=0.1
    )

# Try running it now with eval_dataset
run_inference(0, eval_dataset)

<PIL.PngImagePlugin.PngImageFile image mode=RGB size=824x1280 at 0x7F99A5563BC0>

--- Index 0 ---
Ground Truth: HCG The Specialt in Cancer Care Patient Name: M.R.D. No.: HealthCare Global Enterprises Ltd. #44-45/2, 2nd Cross, Raja Ram Mohan Roy Extn. (Of Lalbagh Double Road) Bangalore-560 027 Tele: +91 80 4020 6109 Fax: +91 80 2248 5962 Thimmareddy Corporate Company Name: Consultant Name: Dranean Ward: R LIPOSOMAL AMPHOTERICIN B INJ. OR. 350 mg x dalys, 4 wks. -INY AMPHOTERICIO - B. 700mg x daily x4 wis OR, INY. ISAVUCONAZOLE Name of the Doctor:: Signature of the Doctorc 200mg 1/1-1-1 x 2 weeks on capsule ISAVUCONAZOLE Loomy 1-0 days Need above medicine please help me if you know any way...
Prediction: HealthCare Global Enterprises Ltd.
# 44-452, 2nd Cross, Raja Ram Mohan Roy Extn. (Off. Lahargh Double Road)
Bangalore - 560 027 Tele. +91 80 4020 6109 Fax: +91 80 2248 5962

Patient Name: Thummarudkar

M.R.D. No.: Corporate Company Name:

Consultant Name: Dr. Awanav Ward:

R


Let's first see before we do any finetuning what the model outputs for the first example!

In [7]:
!pip install jiwer

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 39.9 MB/s eta 0:00:0000:0100:01


<a name="Train"></a>
### Train the model
Now let's train our model. We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

We use our new `UnslothVisionDataCollator` which will help in our vision finetuning setup.

In [8]:
import random
import torch
from jiwer import cer
from tqdm import tqdm

# 1. Setup Selection
num_samples = 10
if len(train_dataset) < num_samples:
    num_samples = len(train_dataset)

indices = random.sample(range(len(train_dataset)), num_samples)

predictions = []
references = []

print(f"Starting inference on {num_samples} random samples...")

# 2. Loop through samples
for idx in tqdm(indices):
    image = train_dataset[idx]["images"][0]
    ground_truth = train_dataset[idx]["messages"][1]["content"][0]["text"]
    instruction = "OCR:"

    messages = [
        {"role": "user", "content": [
            {"type": "image"},
            {"type": "text", "text": instruction}
        ]}
    ]

    text_prompt = processor.tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )

    inputs = processor(
        image, text_prompt, add_special_tokens=False, return_tensors="pt"
    ).to("cuda")

    # Generate
    with torch.inference_mode():
        output_ids = model.generate(
            **inputs,
            max_new_tokens=128,
            use_cache=False,  # <--- CHANGED THIS TO FALSE
            temperature=1.5,
            min_p=0.1
        )

    # Decode Output
    input_length = inputs.input_ids.shape[1]
    generated_ids = output_ids[:, input_length:]
    pred_text = processor.decode(generated_ids[0], skip_special_tokens=True)

    predictions.append(pred_text)
    references.append(ground_truth)

# 3. Calculate CER
score = cer(references, predictions)
print(f"\nResults over {num_samples} samples:")
print(f"Average CER: {score:.4f} (Lower is better)")
print(f"Percentage Accuracy: {(1-score)*100:.2f}%")

Starting inference on 10 random samples...


100%|██████████| 10/10 [15:00<00:00, 90.02s/it] 


Results over 10 samples:
Average CER: 0.5980 (Lower is better)
Percentage Accuracy: 40.20%


In [9]:
from trl import SFTTrainer, SFTConfig
from unsloth.trainer import UnslothVisionDataCollator

FastVisionModel.for_training(model) # Enable for training!

custom_collator = UnslothVisionDataCollator(
    model=model,
    processor=processor,
    ignore_index=-100,
    max_seq_length=1024,
    train_on_responses_only=True,
    instruction_part = "User: ",
    response_part = "\nAssistant:",
    pad_to_multiple_of = 8,
)

trainer = SFTTrainer(
    model = model,
    tokenizer = processor.tokenizer,
    data_collator = custom_collator,
    train_dataset = train_dataset,
    args = SFTConfig(
        per_device_train_batch_size = 2,
        gradient_accumulation_steps = 2, # Use GA to mimic batch size!
        warmup_steps = 5,
       
        num_train_epochs = 3, # Set this instead of max_steps for full training runs
        learning_rate = 5e-6,
        logging_steps = 1,
        optim = "adamw_8bit",
        weight_decay = 0.001,
        lr_scheduler_type = "linear",
        seed = 3407,
        output_dir = "outputs",
        report_to = "none",

        # You MUST put the below items for vision finetuning:
        remove_unused_columns = False,
        dataset_text_field = "",
        dataset_kwargs = {"skip_prepare_dataset": True},
        max_length = 1024,
        fp16 = not torch.cuda.is_bf16_supported(),
        bf16 = torch.cuda.is_bf16_supported(),
    ),
)


In [10]:
# @title Show current memory stats
gpu_stats = torch.cuda.get_device_properties(0)
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)
print(f"GPU = {gpu_stats.name}. Max memory = {max_memory} GB.")
print(f"{start_gpu_memory} GB of memory reserved.")

GPU = Tesla T4. Max memory = 14.741 GB.
6.512 GB of memory reserved.


In [11]:
trainer_stats = trainer.train()


==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 116 | Num Epochs = 3 | Total steps = 87
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 2
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 2 x 1) = 4
 "-____-"     Trainable parameters = 958,588,736 of 958,588,736 (100.00% trained)


Unsloth: Will smartly offload gradients to save VRAM!


In [12]:
# @title Show final memory and time stats
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)
used_percentage = round(used_memory / max_memory * 100, 3)
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)
print(f"{trainer_stats.metrics['train_runtime']} seconds used for training.")
print(
    f"{round(trainer_stats.metrics['train_runtime']/60, 2)} minutes used for training."
)
print(f"Peak reserved memory = {used_memory} GB.")
print(f"Peak reserved memory for training = {used_memory_for_lora} GB.")
print(f"Peak reserved memory % of max memory = {used_percentage} %.")
print(f"Peak reserved memory for training % of max memory = {lora_percentage} %.")

277.3698 seconds used for training.
4.62 minutes used for training.
Peak reserved memory = 13.547 GB.
Peak reserved memory for training = 7.035 GB.
Peak reserved memory % of max memory = 91.9 %.
Peak reserved memory for training % of max memory = 47.724 %.


<a name="Inference"></a>
### Inference
Let's run the model! You can change the instruction and input - leave the output blank!

We use `min_p = 0.1` and `temperature = 1.5`. Read this [Tweet](https://x.com/menhguin/status/1826132708508213629) for more information on why.

In [17]:
import random
import torch
from jiwer import cer
from tqdm import tqdm

# 1. Setup Selection
num_samples = 10
if len(train_dataset) < num_samples:
    num_samples = len(train_dataset)

indices = random.sample(range(len(train_dataset)), num_samples)

predictions = []
references = []

print(f"Starting inference on {num_samples} random samples...")

# 2. Loop through samples
for idx in tqdm(indices):
    image = train_dataset[idx]["images"][0]
    ground_truth = train_dataset[idx]["messages"][1]["content"][0]["text"]
    instruction = "OCR:"

    messages = [
        {"role": "user", "content": [
            {"type": "image"},
            {"type": "text", "text": instruction}
        ]}
    ]

    text_prompt = processor.tokenizer.apply_chat_template(
        messages, tokenize=False, add_generation_prompt=True
    )

    inputs = processor(
        image, text_prompt, add_special_tokens=False, return_tensors="pt"
    ).to("cuda")

    # Generate
    with torch.inference_mode():
        output_ids = model.generate(
            **inputs,
            max_new_tokens=128,
            use_cache=False,  # <--- CHANGED THIS TO FALSE
            temperature=1.5,
            min_p=0.1
        )

    # Decode Output
    input_length = inputs.input_ids.shape[1]
    generated_ids = output_ids[:, input_length:]
    pred_text = processor.decode(generated_ids[0], skip_special_tokens=True)

    predictions.append(pred_text)
    references.append(ground_truth)

# 3. Calculate CER
score = cer(references, predictions)
print(f"\nResults over {num_samples} samples:")
print(f"Average CER: {score:.4f} (Lower is better)")
print(f"Percentage Accuracy: {(1-score)*100:.2f}%")

Starting inference on 10 random samples...


100%|██████████| 10/10 [16:32<00:00, 99.20s/it]


Results over 10 samples:
Average CER: 0.4050 (Lower is better)
Percentage Accuracy: 59.50%


<a name="Save"></a>
### Saving, loading finetuned models
To save the final model as LoRA adapters, either use Huggingface's `push_to_hub` for an online save or `save_pretrained` for a local save.

**[NOTE]** This ONLY saves the LoRA adapters, and not the full model. To save to 16bit or GGUF, scroll down!

In [14]:
model.save_pretrained("lora_model_v3")  # Local saving
tokenizer.save_pretrained("lora_model_v3")
# model.push_to_hub("your_name/lora_model", token = "...") # Online saving
# tokenizer.push_to_hub("your_name/lora_model", token = "...") # Online saving

('lora_model_v3/tokenizer_config.json',
 'lora_model_v3/special_tokens_map.json',
 'lora_model_v3/chat_template.jinja',
 'lora_model_v3/tokenizer.model',
 'lora_model_v3/added_tokens.json',
 'lora_model_v3/tokenizer.json')

Now if you want to load the LoRA adapters we just saved for inference, set `False` to `True`:

### Saving to float16 for VLLM

We also support saving to `float16` directly. Select `merged_16bit` for float16. Use `push_to_hub_merged` to upload to your Hugging Face account! You can go to https://huggingface.co/settings/tokens for your personal tokens.

In [ ]:
# Select ONLY 1 to save! (Both not needed!)

# Save locally to 16bit
if False: model.save_pretrained_merged("unsloth_finetune", tokenizer,)

# To export and save to your Hugging Face account
if False: model.push_to_hub_merged("manojemjay/unsloth_finetune", tokenizer, token = "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")

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️

  This notebook and all Unsloth notebooks are licensed [LGPL-3.0](https://github.com/unslothai/notebooks?tab=LGPL-3.0-1-ov-file#readme)
</div>
